File: Code/0/0/0011/0011.cpp
/******************************************************************************
 * 文件名: 0011
 * 作者: BOLUN XU
 * 创建日期: 2025
 * 版本: 1.0
 * 描述: 盛最多水的容器。
 * 时间复杂度： O(N)
 * 空间复杂度： O(1)
 ******************************************************************************/
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

class Solution {
public:
  int maxArea(vector<int>& height) {
    int left = 0;
    int right = height.size() - 1;
    int max_area = 0;
    while (left < right) {
      if (height[left] < height[right]) {
        int area = height[left] * (right - left);
        max_area = max(max_area, area);
        left++;
      }
      else {
        int area = height[right] * (right - left);
        max_area = max(max_area, area);
        right--;
      }
    }
    return max_area;
  }
};

int main() {
  Solution solution;
  vector<int> height = {1, 8, 6, 2, 5, 4, 8, 3, 7};
  cout << solution.maxArea(height) << endl;
}

File: Code/0/0/0011/0011.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
文件名: PersonalLeetCode 0011.py
作者: Bolun Xu
创建日期: 2025/3/25
版本: 2.0
描述: 盛最多水的容器。
时间复杂度： O(N)
空间复杂度： O(1)
"""
from typing import List


class Solution:
    def maxArea(self, height: List[int]) -> int:
        left = 0
        right = len(height) - 1
        max_area = 0
        while left < right:
            if height[left] < height[right]:
                area = height[left] * (right - left)
                max_area = max(area, max_area)
                left += 1
            else:
                area = height[right] * (right - left)
                max_area = max(area, max_area)
                right -= 1
        return max_area


if __name__ == '__main__':
    solution = Solution()
    print(solution.maxArea([1, 8, 6, 2, 5, 4, 8, 3, 7]))
